ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/LedgerPeers.hs

```diff
@@ -161,6 +172,122 @@ instance Arbitrary ArbLedgerPeersKind where
+                       \(LedgerPools lp, sp, ArbLedgerPeersKind lpk, rng') -> do
+                    threadDelay 1900 -- we need to invalidate ledger peer's cache
+                    let maxRequest = case sp of
+                                       Just (LedgerPools sp') -> min (length lp) (length sp')
+                                       Nothing -> length lp
+                        numRequested = NumberOfPeers . fromIntegral . fst . randomR (1, maxRequest) $ rng'
+                    resp <- request numRequested lpk
+                    pure $ case resp of
+                      Nothing          -> (numRequested, Set.empty)
+                      Just (peers, _)  -> (numRequested, Set.fromList [ RelayAccessAddress ip port
+                                                                      | Just (ip, port) <-     IP.fromSockAddr
+                                                                                           <$> Set.toList peers]))
+
+  return . ioProperty $ do
+    tr' <- evaluateTrace (runSimTrace sim)
+    case tr' of
+      SimException e trace -> do
+        return $ counterexample (intercalate "\n" $ show e : trace) False
+      SimDeadLock trace -> do
+        return $ counterexample (intercalate "\n" $ "Deadlock" : trace) False
+      SimReturn results _trace -> do
+        return $ either (`counterexample` False) id evalChecks
+        where
+          evalChecks = foldrM step (property True) (zip6 results snapshotSlots' ledgerSlots ledgerPeersKinds ledgerPools snapshotPools')
+
+          step ((numRequested, peers), ss, ls, ArbLedgerPeersKind lpk, lpool, spool) pass =
+            let fixupSnapshotSlot = maybe (-1) getPositive (ss <$ spool)
+                snapshotSet spools = Set.fromList (concatMap (NonEmpty.toList . snd) $ getLedgerPools spools)
+                ledgerSet = Set.fromList (concatMap (NonEmpty.toList . snd) $ getLedgerPools lpool)
+                bigLedgerSet = Set.fromList (concatMap (NonEmpty.toList . snd . snd) . accumulateBigLedgerStake . getLedgerPools $ lpool)
+                source =
+                  case spool of
+                    Just spools | fixupSnapshotSlot >= getPositive ls  ->
+                                  snapshotSet spools `Set.union` ledgerSet
+                    _otherwise -> case lpk of
+                                    AllLedgerPeers -> ledgerSet
+                                    BigLedgerPeers -> bigLedgerSet
+            in if peers `Set.isSubsetOf` source
+               then return pass
+               else Left $ intercalate "\n"
+                             ["Counterexample:", "Requested: "++ show numRequested,
+                              "Type requested: " ++ show lpk,
+                              "Ledger slot: " ++ show ls, "Snapshot slot? " ++ show (ss <$ spool),
+                              show peers, "========================", "violates Set.isSubsetOf:",
+                              "========================", show source]
```

Comment:
Why is this test so complicated? Why don't you factor out the request logic and then just test that single function instead of running a simulation?

Reply:
Done